compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/ReachabilityAnalyzer.java

```diff
@@ -184,14 +216,13 @@ public void visit(BLangSimpleVariableDef varDefNode) {
         checkStatementExecutionValidity(varDefNode);
 
         BLangExpression expr = varDefNode.var.expr;
-        if (expr != null && expr.getKind() == NodeKind.DO_ACTION) {
-            analyzeReachability(expr, env);
-        }
+        analyzeReachabilityInExpressionIfApplicable(expr, env);
     }
 
     @Override
     public void visit(BLangAssignment assignNode) {
         checkStatementExecutionValidity(assignNode);
+        analyzeReachabilityInExpressionIfApplicable(assignNode.expr, env);
```

Comment:
Need to handle destructuring assignment statements also. The following doesn't result in an error atm.

```ballerina
public function main() returns error? {
    var error(m) = <error> from var item in 1 ... 5
        where false
        do {
            int _ = 1;
        };
}
```

Reply:
Fixed